src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -12,100 +12,473 @@
 
 'use strict';
 
-type UpdateQueueNode = {
-  partialState: any,
-  callback: ?Function,
+import type { Fiber } from 'ReactFiber';
+import type { PriorityLevel } from 'ReactPriorityLevel';
+
+const {
+  ForceUpdate,
+  Callback: CallbackEffect,
+} = require('ReactTypeOfSideEffect');
+
+const {
+  NoWork,
+  SynchronousPriority,
+  TaskPriority,
+} = require('ReactPriorityLevel');
+
+type PartialState<State, Props> =
+  $Subtype<State> |
+  (prevState: State, props: Props) => $Subtype<State>;
+
+type Callback = () => void;
+
+type Update = {
+  priorityLevel: PriorityLevel,
+  partialState: PartialState<any, any>,
+  callback: Callback | null,
   isReplace: boolean,
-  next: ?UpdateQueueNode,
+  isForced: boolean,
+  next: Update | null,
 };
 
-export type UpdateQueue = UpdateQueueNode & {
-  isForced: boolean,
-  hasUpdate: boolean,
-  hasCallback: boolean,
-  tail: UpdateQueueNode
+// Singly linked-list of updates. When an update is scheduled, it is added to
+// the queue of the current fiber and the work-in-progress fiber. The two queues
+// are separate but they share a persistent structure.
+//
+// During reconciliation, updates are removed from the work-in-progress fiber,
+// but they remain on the current fiber. That ensures that if a work-in-progress
+// is aborted, the aborted updates are recovered by cloning from current.
+//
+// The work-in-progress queue is always a subset of the current queue.
+//
+// When the tree is committed, the work-in-progress becomes the current.
+export type UpdateQueue = {
+  first: Update | null,
+  last: Update | null,
+
+  // Dev only
+  isProcessing?: boolean,
 };
 
-exports.createUpdateQueue = function(partialState : mixed) : UpdateQueue {
-  const queue = {
+function comparePriority(a : PriorityLevel, b : PriorityLevel) : number {
+  // When comparing update priorities, treat sync and Task work as equal.
+  // TODO: Could we avoid the need for this by always coercing sync priority
+  // to Task when scheduling an update?
+  if ((a === TaskPriority || a === SynchronousPriority) &&
+      (b === TaskPriority || b === SynchronousPriority)) {
+    return 0;
+  }
+  if (a === NoWork && b !== NoWork) {
+    return -Infinity;
```

Comment:
Oh nvm I thought this was a different line (on my phone). Okay I'll just use like 999 or something :D